kroxylicious/src/main/java/io/kroxylicious/proxy/config/PluginConfigTypeIdResolver.java

```diff
@@ -3,41 +3,48 @@
  *
  * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
  */
+
 package io.kroxylicious.proxy.config;
 
+import java.util.Map;
+
 import com.fasterxml.jackson.annotation.JsonTypeInfo;
 import com.fasterxml.jackson.databind.DatabindContext;
 import com.fasterxml.jackson.databind.JavaType;
 import com.fasterxml.jackson.databind.jsontype.impl.TypeIdResolverBase;
 
-import io.kroxylicious.proxy.service.FilterFactoryManager;
+public class PluginConfigTypeIdResolver extends TypeIdResolverBase {
 
-public class FilterConfigTypeIdResolver extends TypeIdResolverBase {
+    private final Map<String, ServiceBasedPluginFactoryRegistry.ProviderAndConfigType<?>> providers;
     private JavaType superType;
 
+    PluginConfigTypeIdResolver(Map<String, ServiceBasedPluginFactoryRegistry.ProviderAndConfigType<?>> providers) {
+        this.providers = providers;
+    }
+
     @Override
     public void init(JavaType baseType) {
         superType = baseType;
     }
 
     @Override
     public String idFromValue(Object value) {
-        throw new UnsupportedOperationException("Not implemented");
+        throw new UnsupportedOperationException();
     }
 
     @Override
     public String idFromValueAndType(Object value, Class<?> suggestedType) {
-        throw new UnsupportedOperationException("Not implemented");
+        throw new UnsupportedOperationException();
     }
 
     @Override
     public JsonTypeInfo.Id getMechanism() {
-        throw new UnsupportedOperationException("Not implemented");
+        throw new UnsupportedOperationException();
     }
 
     @Override
     public JavaType typeFromId(DatabindContext context, String id) {
-        Class<?> subType = FilterFactoryManager.INSTANCE.getConfigType(id);
-        return context.constructSpecializedType(superType, subType);
+        var provider = providers.get(id);
+        return context.constructSpecializedType(superType, provider.config());
```

Comment:
If you make a mistake in the configuration and refer to a non-existent type this results in a NPE being thrown on Kroxy start which should probably be something more readable talking about the failed id.

```
2023-10-17 22:37:40 <main> ERROR io.kroxylicious.proxy.StartupShutdownLogger:71 - Exception on startup
java.lang.IllegalArgumentException: Couldn't parse configuration
	at io.kroxylicious.proxy.config.ConfigParser.parseConfiguration(ConfigParser.java:73) ~[kroxylicious-0.3.0-SNAPSHOT.jar:?]
	at io.kroxylicious.app.Kroxylicious.call(Kroxylicious.java:63) [kroxylicious-app-0.3.0-SNAPSHOT.jar:?]
	at io.kroxylicious.app.Kroxylicious.call(Kroxylicious.java:33) [kroxylicious-app-0.3.0-SNAPSHOT.jar:?]
	... snip
	at io.kroxylicious.app.Kroxylicious.main(Kroxylicious.java:91) [kroxylicious-app-0.3.0-SNAPSHOT.jar:?]
Caused by: com.fasterxml.jackson.databind.exc.ValueInstantiationException: Cannot construct instance of `io.kroxylicious.proxy.config.FilterDefinition`, problem: Cannot invoke "io.kroxylicious.proxy.config.ServiceBasedPluginFactoryRegistry$ProviderAndConfigType.config()" because "provider" is null
 at [Source: (sun.nio.ch.ChannelInputStream); line: 3, column: 1] (through reference chain: io.kroxylicious.proxy.config.Configuration["filters"]->java.util.ArrayList[0])
	at com.fasterxml.jackson.databind.exc.ValueInstantiationException.from(ValueInstantiationException.java:47) ~[jackson-databind-2.15.3.jar:2.15.3]
	... snip
	at io.kroxylicious.proxy.config.ConfigParser.parseConfiguration(ConfigParser.java:70) ~[kroxylicious-0.3.0-SNAPSHOT.jar:?]
	... 11 more
Caused by: java.lang.NullPointerException: Cannot invoke "io.kroxylicious.proxy.config.ServiceBasedPluginFactoryRegistry$ProviderAndConfigType.config()" because "provider" is null
	at io.kroxylicious.proxy.config.PluginConfigTypeIdResolver.typeFromId(PluginConfigTypeIdResolver.java:48) ~[kroxylicious-0.3.0-SNAPSHOT.jar:?]
	... snip
```

Reply:
Improved. We should probably have a dedicated exception type for bad configuration, but that's for another PR.